File: DatabaseInteract.java
import com.google.api.core.ApiFuture;
import com.google.cloud.firestore.DocumentReference;
// [START fs_include_dependencies]
import com.google.cloud.firestore.Firestore;
import com.google.cloud.firestore.FirestoreOptions;
// [END fs_include_dependencies]
import com.google.cloud.firestore.QueryDocumentSnapshot;
import com.google.cloud.firestore.QuerySnapshot;
import com.google.cloud.firestore.WriteResult;
import com.google.common.collect.ImmutableMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class DatabaseInteract {
	
	// Use the application default credentials
	GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
	FirebaseOptions options = new FirebaseOptions.Builder()
	    .setCredentials(credentials)
	    .setProjectId(projectId)
	    .build();
	FirebaseApp.initializeApp(options);

	Firestore db = FirestoreClient.getFirestore();
	
	public String getMsgId() {
		DocumentReference docRef = db.collection("message").document("msgid");
		
		ApiFuture<DocumentSnapshot> future = docRef.get();	
		DocumentSnapshot document = future.get();
		
		if (document.exists()) {
		  System.out.println("Document data: " + document.getData());
		  Map<String, Object> map = document.getData();
		  return map.get("msgid");
		} else {
		  System.out.println("No such document!");
		  return null;
		}
    }
	
	public void nextMsgId() {
		int msgid = Integer.parseInt(getMsgId()) + 1;
		
		// Create a Map to store the data we want to set
		Map<String, Object> docData = new HashMap<>();
		docData.put("msgid", msgid);
		
		// Add a new document (asynchronously) in collection "cities" with id "LA"
		ApiFuture<WriteResult> future = db.collection("message").document("msgid").set(docData);
		// ...
		// future.get() blocks on response
		System.out.println("Update time : " + future.get().getUpdateTime());
	}
	
	public void addUser(User user) {
		// Create a Map to store the data we want to set
			Map<String, Object> docData = new HashMap<>();
			docData.put("username", user.username);
			docData.put("password", user.password);
				
		// Add a new document (asynchronously) in collection "cities" with id "LA"
		ApiFuture<WriteResult> future = db.collection("user").document(username).set(docData);
	}
	
	public User queryUser(String username) {
		DocumentReference docRef = db.collection("user").document(username);
		// asynchronously retrieve the document
		ApiFuture<DocumentSnapshot> future = docRef.get();
		// block on response
		DocumentSnapshot document = future.get();
		User user = null;
		if (document.exists()) {
		  // convert document to POJO
		  user = document.toObject(User.class);
		  System.out.println(user);
		} else {
		  System.out.println("No such document!");
		}
		return user;
	}
	
	public void addMessage (Message msg) {
		// Create a Map to store the data we want to set
			Map<String, Object> docData = new HashMap<>();
			docData.put("id", getMsgId() + 1);
			docData.put("textMessage", msg.textMessage);
			docData.put("audioMessage", msg.audioMessage);
			docData.put("pictureMessage", msg.pictureMessage);
			docData.put("locationx", msg.locationx);
			docData.put("locationy", msg.locationy);
			docData.put("timestamp", msg.timestamp);
			docData.put("flaggedCount", 0);
			
			ApiFuture<WriteResult> future = db.collection("message").document((getMsgId()+1).toString()).set(docData);
			nextMsgId();
	}
	
	public Message queryMessage(int msgid) {
		DocumentReference docRef = db.collection("message").document(msgid.toString());
		// asynchronously retrieve the document
		ApiFuture<DocumentSnapshot> future = docRef.get();
		// block on response
		DocumentSnapshot document = future.get();
		Message msg = null;
		if (document.exists()) {
		  // convert document to POJO
		  msg = document.toObject(Message.class);
		  System.out.println(msg);
		} else {
		  System.out.println("No such document!");
		}
		return msg;
	}
	
	public void deleteMessage(int msgid) {
		// asynchronously delete a document
		ApiFuture<WriteResult> writeResult = db.collection("message").document(msgid.toString()).delete();
		// ...
		System.out.println("Update time : " + writeResult.get().getUpdateTime());
	}
	
	public ArrayList<User> getUser () {
		ArrayList<User> res = new ArrayList<User> ();
		//asynchronously retrieve all documents
		ApiFuture<QuerySnapshot> future = db.collection("user").get();
		// future.get() blocks on response
		List<QueryDocumentSnapshot> documents = future.get().getDocuments();
		for (QueryDocumentSnapshot document : documents) {
			res.push(document.toObject(User.class));
		}
		return res;
	}
	
	public ArrayList<Message> getMessage () {
		ArrayList<Message> res = new ArrayList<Message> ();
		//asynchronously retrieve all documents
		ApiFuture<QuerySnapshot> future = db.collection("message").get();
		// future.get() blocks on response
		List<QueryDocumentSnapshot> documents = future.get().getDocuments();
		for (QueryDocumentSnapshot document : documents) {
			res.push(document.toObject(Message.class));
		}
		return res;
	}
}

File: databaseInteract.swift
//Reference database as db elsewhere AppRef
import Firebase


class DatabaseInteract {
    var db: Firestore!
    
    func getMsgId() -> Int {
        let docRef = db.collection("message").document("msgid")
        var id = Int()
        docRef.getDocument { (docSnapShot, error) in
            guard let docSnapshot = docSnapShot, docSnapshot.exists else {return }
            let myData = docSnapshot.data()
            id = myData?["msgid"] as? Int ?? 0
            
        }
        return id
    }

    func nextMsgId() {
        //proceed to next message id once new msg is added

        db.collection("id").document("msgid").setData([
            "id": getMsgId()+1
        ]) { err in
            if let err = err {  
                print("Error writing document: \(err)")
            } else {
                print("Document successfully written!")
            }
        }
    }

    
    func addUser(user: User){
        //Add a new document named username into user
        
        db.collection("user").document(user.username).setData([
            "username": user.username,
            "password": user.password,
            "userid": 1
        ]) { err in
            if let err = err {  
                print("Error writing document: \(err)")
            } else {
                print("Document successfully written!")
            }
        }
    }

    func queryUser (username: String) -> User {
        //returns User object that has same username
        let docRef = db.collection("user").document(username)
        let user = User()

        docRef.getDocument { (docSnapShot, error) in
            guard let docSnapshot = docSnapShot, docSnapshot.exists else {return }
            let myData = docSnapshot.data()
            let username = myData?["username"] as? String ?? ""
            let password = myData?["password"] as? String ?? ""
            user.setUsername(un: username)
            user.setPassword(pw: password)
        }
        return user
    }

    func addMessage (msg: Message){
        //Add a new document named message id into message collection
        db.collection("message").document(String(msg.getID())).setData([
            "id": getMsgId()+1,
            "textMessage": msg.textMessage,
            "audioMessage": msg.audioMessage,
            "pictureMessage": msg.pictureMessage,
            "latitude": msg.latitude,
            "longitude": msg.longitude,
            "timestamp": msg.timestamp,
            "flaggedCount": 0
        ]) { err in
            if let err = err {  
                print("Error writing document: \(err)")
            } else {
                print("Document successfully written!")
                self.nextMsgId();
            }
        }

    }
    
    func queryMessage (msgid: Int) -> Message {
        //returns Message object w msgid
        let docRef = db.collection("message").document("msgid")
        let message = Message()
        
        docRef.getDocument { (docSnapShot, error) in
            guard let docSnapshot = docSnapShot, docSnapshot.exists else {return }
            let myData = docSnapshot.data()
            let id = myData?["id"] as? Int ?? 0
            let pictureMessage = myData?["pictureMessage"] as? String ?? ""
            let audioMessage = myData?["audioMessage"] as? String ?? ""
            let textMessage = myData?["textMessage"] as? String ?? ""
            let flaggedCount = myData?["flaggedCount"] as? Int ?? 0
            let timestamp = myData?["timestamp"] as? Date ?? Date()
            let latitude = myData?["latitude"] as? Double ?? 0.0
            let longitude = myData?["longitude"] as? Double ?? 0.0
    
            message.setID(id: id)
            message.setImage(picture: pictureMessage)
            message.setAudio(audio: audioMessage)
            message.setText(text: textMessage)
            message.setFlaggedCount(flagged: flaggedCount)
            message.setTimestamp(time: timestamp)
            message.setLatitude(lat: latitude)
            message.setLongitude(long: longitude)
        }
        return message
    }

    func deleteMessage (msgid: Int){
        db.collection("message").document("msgid").delete() { err in
            if let err = err {
                print("Error removing document: \(err)")
            } else {
                print("Document successfully removed!")
            }
        }
    }
/*
    func getUsers() -> String{
        //return entire user collection
        db.collection("user").getDocuments() { (querySnapshot, err) in
            if let err = err {
                print("Error getting documents: \(err)")
            } else {
                return document.date()
            }
        }
    }

    func getMessages() {
        //return entire message collection
        db.collection("message").getDocuments() { (querySnapshot, err) in
            if let err = err {
                print("Error getting documents: \(err)")
            } else {
                return document.date()
            }
        }
    }
     */

}
